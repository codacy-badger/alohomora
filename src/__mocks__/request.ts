import memberIDResponse from "./_slack_get_member_id_response.json";
import imOpenResponse from "./_slack_post_im_open_response.json";
import messageResponse from "./_slack_post_message_response.json";

const RESPONSE_MAP = {
  "users.lookupByEmail": memberIDResponse,
  "im.open": imOpenResponse,
  "chat.postMessage": messageResponse
};

function getResponseForURL(url: string) {
  const api = Object.keys(RESPONSE_MAP).find(key => url.indexOf(key) !== -1);

  return api && RESPONSE_MAP[api as keyof typeof RESPONSE_MAP];
}

const request = jest.fn((options, callback) => {
  const { url } = options;
  const { response = {}, content = {} } = getResponseForURL(url) || {};

  process.nextTick(() => callback(null, response, content));
});

export default request;
