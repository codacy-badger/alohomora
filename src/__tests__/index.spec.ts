import request from "request";
import lambda from "../index";

const requestMock = (request as unknown) as jest.Mock;

const event = {
  to: "testing-gago-channel",
  message: "some content to *show* how _amazing_ this ~is~"
};
describe("on Post on Slack", () => {
  beforeEach(() => {
    requestMock.mockClear();
  });
  it("should send proper default arguments", async () => {
    await lambda(event);
    const slackRequest = requestMock.mock.calls.find(([{ url }]) =>
      url.endsWith("chat.postMessage")
    );
    expect(slackRequest[0].body.parse).toEqual("full");
    expect(slackRequest[0]).toMatchSnapshot();
  });

  it("should post to a channel with the right parameters", () => {
    return expect(lambda(event)).resolves.toMatchSnapshot();
  });

  it("should fail with the wrong parameters", () => {
    return expect(lambda({})).rejects.toMatchSnapshot();
  });

  it("should post a DM with the right parameters", () => {
    return expect(
      lambda({ ...event, to: "gago@coursera.org" })
    ).resolves.toMatchSnapshot();
  });

  it("allows overriding of the `parse` parameter", async () => {
    await lambda({ ...event, to: "gago@coursera.org", parse: "none" });
    const slackRequest = requestMock.mock.calls.find(([{ url }]) =>
      url.endsWith("chat.postMessage")
    );
    expect(slackRequest[0].body.parse).toEqual("none");
    expect(slackRequest[0]).toMatchSnapshot();
  });
});
