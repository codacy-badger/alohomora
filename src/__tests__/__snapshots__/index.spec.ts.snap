// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on Post on Slack allows overriding of the \`parse\` parameter 1`] = `
Object {
  "auth": Object {
    "bearer": "TOKEN",
  },
  "body": Object {
    "attachments": Array [],
    "channel": "DBB7G0WBG",
    "icon_emoji": ":speaking_head_in_silhouette:",
    "link_names": true,
    "parse": "none",
    "text": "some content to *show* how _amazing_ this ~is~",
    "username": undefined,
  },
  "json": true,
  "method": "POST",
  "url": "https://slack.com/api/chat.postMessage",
}
`;

exports[`on Post on Slack should fail with the wrong parameters 1`] = `[Error: channel and/or message missing]`;

exports[`on Post on Slack should post a DM with the right parameters 1`] = `
Object {
  "channel": "CB6S4V8NP",
  "message": Object {
    "attachments": Array [
      Object {
        "fallback": "And here's an attachment!",
        "id": 1,
        "text": "And here's an attachment!",
      },
    ],
    "bot_id": "BB5KBR5Q8",
    "subtype": "bot_message",
    "text": "some content to *show* how _amazing_ this ~is~",
    "ts": "1528914636.000177",
    "type": "message",
    "username": "bot",
  },
  "ok": true,
  "response_metadata": Object {
    "warnings": Array [
      "missing_charset",
    ],
  },
  "ts": "1528914636.000177",
  "warning": "missing_charset",
}
`;

exports[`on Post on Slack should post to a channel with the right parameters 1`] = `
Object {
  "channel": "CB6S4V8NP",
  "message": Object {
    "attachments": Array [
      Object {
        "fallback": "And here's an attachment!",
        "id": 1,
        "text": "And here's an attachment!",
      },
    ],
    "bot_id": "BB5KBR5Q8",
    "subtype": "bot_message",
    "text": "some content to *show* how _amazing_ this ~is~",
    "ts": "1528914636.000177",
    "type": "message",
    "username": "bot",
  },
  "ok": true,
  "response_metadata": Object {
    "warnings": Array [
      "missing_charset",
    ],
  },
  "ts": "1528914636.000177",
  "warning": "missing_charset",
}
`;

exports[`on Post on Slack should send proper default arguments 1`] = `
Object {
  "auth": Object {
    "bearer": "TOKEN",
  },
  "body": Object {
    "attachments": Array [],
    "channel": "testing-gago-channel",
    "icon_emoji": ":speaking_head_in_silhouette:",
    "link_names": true,
    "parse": "full",
    "text": "some content to *show* how _amazing_ this ~is~",
    "username": undefined,
  },
  "json": true,
  "method": "POST",
  "url": "https://slack.com/api/chat.postMessage",
}
`;
